networking_ccloud/ml2/agent/common/switch.py

```diff
@@ -26,20 +28,67 @@
 LOG = logging.getLogger(__name__)
 
 
-def run_in_executor(type_):
+class FullSyncScheduled(Exception):
+    """Raised if a scheduled write-attempt was voluntarily interrupted because there's a full-sync in the queue
+
+    Callers catching this exception should wait for the attached future
+    attribute instead to know when their task finishes.
+    """
+    def __init__(self, *, future):
+        super().__init__(self)
+        self.future = future
+
+
+def run_in_executor(type_, replacable_by_full_sync=False):
     """Decorator to run a method in the ThreadPoolExecutor of the class
 
     The wrapped method thus returns a futurist.Future now instead of its actual
     result.
+
+    If replacable_by_full_sync is set, the decorated method will not run if
+    there's a full-sync scheduled. Instead, it will raise FullSyncScheduled.
+    Additionally, the decorated method will also not get scheduled, if there's
+    already a full-sync scheduled. In that case, the wrapped method returns the
+    scheduled full-sync's futurist.Future instead of its own. Callers have to
+    be able to handle that.
     """
+    if type_ != 'write' and replacable_by_full_sync:
+        raise ValueError(f"@run_in_executor() called with replacable_by_full_sync=True on {type_} executor - "
+                         "only 'write' is supported.")
+
     def decorator(fn):
 
         @wraps(fn)
         def wrapped(self, *args, **kwargs):
+            executor = getattr(self, f'_{type_}_executor')
+
             current_context = context.get_current()
 
+            if replacable_by_full_sync:
+                # NOTE(jkulik): This is a very rough algorithm as starting a
+                # point and up for discussion. It assumes that a full-sync is
+                # a better/faster solution than running all the tasks in the
+                # queue. It only holds true for tasks that are
+                # replacable_by_full_sync, but the assumption is that these are
+                # most of the tasks. Average runtime is skewed by having the
+                # full-sync run in the same queue.
+                avg_runtime = executor.statistics.average_runtime if executor.statistics.executed else 0
+                qsize = executor._work_queue.qsize()
+                full_sync_threshold = 120
+                if qsize * avg_runtime > full_sync_threshold:
+                    LOG.warning('Scheduling full-sync for queue_size %s and avg runtime %s as the threshold of %s is '
+                                'reached', qsize, avg_runtime, full_sync_threshold)
```

Comment:
Could you also ad the switch name here, so we know which switch exceeded the size? Btw, printing `self` should do the needful there, afair it has a proper `__str__()` that prints out name and ip.